ESQL: log query and execution time

alex-spies opened the pull request:
Log the user's query and its execution time, both in case of success and in case of failure.

alex-spies committed:
Log successful queries and execution time

alex-spies committed:
Log queries and exec times also on failure

alex-spies committed:
Update docs/changelog/99058.yaml

costin commented:
1. Perform the logging at the REST level (RestEsqlQueryAction) since that's what the users sees.
2 The ActionListener only needs to perform logging - thus there shouldn't be any extra error checking/catching.

alex-spies committed:
Log query execution time in EsqlResponseListener

alex-spies committed:
Add comment

alex-spies committed:
Add logging in case of failed execution back in

alex-spies commented:
1. @costin I moved the logging of successful queries into `EsqlResponseLilstener.java`, after the REST response is assembled. I agree that this is more accurate.
2. Unfortunately, the catch-log-rethrow block that I added is necessary if we want to log _all_ incoming ESQL queries; in case of an exception during planning or during parsing of the actual ESQL query, the error handling of the action listener is not triggered. This can be seen even in queries with trivial typos like `{"query": "row a = 1 \n | keepa"}`. I can remove the catch-log-rethrow block, but IMO that would make our logging of queries incomplete esp. because we won't have logs of queries that lead to exceptions during query planning.

alex-spies committed:
Remove unnecessary comment

alex-spies commented:
@elasticsearchmachine run elasticsearch-ci/cloud-deploy

alex-spies committed:
Merge branch 'main' into esql-log-queries

alex-spies commented:
Alright, made a cloud deployment based on this PR and can confirm that the logs nicely make their way into the admin console:
![image](https://github.com/elastic/elasticsearch/assets/88324834/1d717848-e80e-40ef-97ba-0bc1a567751e)

alex-spies commented:
@elasticsearchmachine run elasticsearch-ci/packaging-tests-unix-sample

alex-spies commented:
@elasticsearchmachine run elasticsearch-ci/packaging-tests-windows-sample

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

alex-spies committed:
Use logging listener in TransportEsqlQueryAction

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlResponseListener.java:
```diff
@@ -27,14 +30,21 @@ public class EsqlResponseListener extends RestResponseListener<EsqlQueryResponse
     private final MediaType mediaType;
+    /**
+     * Keep the initial query for logging purposes.
+      */
```
One extra space here. Does the formatter hate this?

alex-spies replied:
Weird, doesn't seem to bother the formatter. I'll fix it manually.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlResponseListener.java:
```diff
@@ -27,14 +30,21 @@ public class EsqlResponseListener extends RestResponseListener<EsqlQueryResponse
     private final long startNanos = System.nanoTime();
     private static final String HEADER_NAME_TOOK_NANOS = "Took-nanos";
 
+    private static final Logger LOGGER = LogManager.getLogger(EsqlResponseListener.class);
```
It's best to stick all of the `static` things above the non-static members. Or, at least, it's more traditional. That way you can scan the list of know what goes where.

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-log-queries

alex-spies committed:
Improve formatting

alex-spies commented:
@elasticsearchmachine run elasticsearch-ci/packaging-tests-unix-sample

costin commented:
> 2\. in case of an exception during planning or during parsing of the actual ESQL query, the error handling of the action listener is not triggered

That looks like a bug - namely parsing is done without wrapping the method into the listener which causes the exception to be thrown without bubbling it up through the listener. 
It's line `optimizedPhysicalPlan(parse(request.query(), request.params()), listener)` --> look at the nested methods and see that each one relies on another to create its input:
optimizedPhysicalPlan is called using a logicalPlan and does the conversion by calling physicalPlan which calls optimizedPlan and so on. It should go all the way to the top and accept the input as a string but the pattern is not applied for parsing.

Once that's improved, there should be no need for a try/catch.

dnhatn commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/TransportEsqlQueryAction.java:
```diff
@@ -94,20 +100,37 @@ private void doExecuteForked(Task task, EsqlQueryRequest request, ActionListener
+                    )
+                )
+            );
+        } catch (RuntimeException e) {
```
We should not catch the exception here; I don't think it will take effect. Instead of simply delegating the failure with `delegateFailureAndWrap`, I think we should wrap both the onResponse and onFailure methods of a listener.

alex-spies replied:
I am not sure if I understand correctly; in the current state of this PR, this catch block is necessary because I otherwise cannot log the query if it failed during parsing. I confirmed locally and on cloud deployment that this catch-block takes effect and logs the query in case of parsing errors.

I want to push a change though, which will remove this catch block and instead ensure that query parsing errors are correctly sent up the listeners rather than catching them separately, as Costin suggested.

alex-spies commented:
> > 2. in case of an exception during planning or during parsing of the actual ESQL query, the error handling of the action listener is not triggered
> 
> That looks like a bug - namely parsing is done without wrapping the method into the listener which causes the exception to be thrown without bubbling it up through the listener. It's line `optimizedPhysicalPlan(parse(request.query(), request.params()), listener)` --> look at the nested methods and see that each one relies on another to create its input: optimizedPhysicalPlan is called using a logicalPlan and does the conversion by calling physicalPlan which calls optimizedPlan and so on. It should go all the way to the top and accept the input as a string but the pattern is not applied for parsing.
> 
> Once that's improved, there should be no need for a try/catch.

Alright, I'll move the parsing so that exceptions during parsing are properly bubbled up and handled on a higher level; then I can get rid of the catch-log-rethrow logic and everything should be logged nicely.

alex-spies committed:
Log by only wrapping EsqlResponseListener

EsqlResponseListener assembles the final REST response and handles all
exceptions thrown during execution, including parsing. Wrapping it
enables logging of everything in one place, only.

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-log-queries

alex-spies commented:
@costin and @dnhatn , I think both your remarks are addressed now.

It wasn't obvious to me at first, but in [RestEsqlQueryAction::prepareRequest](https://github.com/elastic/elasticsearch/blob/8752d8045d504ea7ef7f7572ea4163d3fa7d4211/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/RestEsqlQueryAction.java#L50), calling `cancellableClient.execute(...)` delegates all failures from execution to the `EsqlResponseListener` that is passed as argument. This includes parsing failures, so there's no need to wrap the esql query parsing in a listener layer.

Wrapping the `EsqlResponseListener` allows to nicely log successes and failures in one place (and using only one timing mechanism). The only downside is that the logged execution time _is slightly higher_ than the one that is sent to the user, because now the log is done after the response is sent (I think).

alex-spies committed:
Use ESQL instead of ES|QL in logs

This improves searchability when viewing the logs in Kibana.

alex-spies commented:
By the way, I think we should use `ESQL` in logs rather than `ES|QL`. When searched via Kibana, `ES|QL` would require a custom analyzer to be searched efficiently, otherwise this becomes 2 separate terms `ES` and `QL`.
![Pasted image 20230906105302](https://github.com/elastic/elasticsearch/assets/88324834/7441ad32-064c-49f2-b84b-6592321a08c2)

I pushed a change to only use `ESQL` in the logs added in this PR.

costin reviewed (approved):
LGTM

dnhatn commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlResponseListener.java:
```diff
@@ -69,8 +78,28 @@ public RestResponse buildResponse(EsqlQueryResponse esqlResponse) throws Excepti
+        // We need to measure the execution time after handling the response/failure for the measurement to be correct.
+        return ActionListener.wrap(r -> {
+            onResponse(r);
+            LOGGER.info("Successfully executed ESQL query in {}ms:\n{}", timeMillis(), esqlQuery);
```
I believe the execution time should exclude the time it takes to send the REST response. Hence, we should log before onResponse() or onFailure().

alex-spies replied:
Logging before `onFailure` is probably fine, but logging before `onResponse` will exclude the time to create the actual REST response (which could take a while depending on result set size).

I can, however, make it so that we log the same time that we also send back to the user [in the response header](https://github.com/elastic/elasticsearch/blob/386103f7b8ba8b454b349b057cad7205434de457/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlResponseListener.java#L82).

I'd like to create a follow-up PR for this to add the more accurate logging in a clean way. (I could just move the success logging to the same location where the response time header is set, but I'd prefer to keep the success/failure logging in one place since the functionalities are related.)

dnhatn reviewed (approved):
I left a comment, but LGTM. Thanks @alex-spies.
